### PR TITLE
Add `plutus-list`

### LINKED ISSUE #81: Add `PlutusTx.Natural.length :: Foldable t => t a -> Natural`
The definition is trivial, but this is the place for it.

### PR DESCRIPTION
**Pull request by re-xyr**
WIP; Will fix #81.

- ~~Do we need this separate module? It may be better to just add the functions into `plutus-extra`.~~ (cf. [comment (#81)](https://github.com/Liqwid-Labs/plutus-extra/issues/81#issuecomment-974692109))
- ~~`ordNub` requires either `sort` or a `Set`, and none is in PlutusTx; so it is not doable for now. Or I missed something?~~
- README is currently empty. Will fill something in
- ~~Do the property tests look OK?~~

### CODE REVIEW

File: plutus-list/plutus-list.cabal
Comment (kozross) [inline plutus-list/plutus-list.cabal:60]: This should be named something slightly more meaningful, since we're likely to have multiple test suites. If this is a suite of property tests, `properties` is fine.

File: plutus-list/plutus-list.cabal
Comment (kozross) [inline plutus-list/plutus-list.cabal:63]: Better name it `Main.hs`, as this makes it clear it's executable.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Natural.hs:8]: This should be your name and contact details: after all, you wrote it!

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Natural.hs:26]: I'm in favour of the 'Haddock, then INLINEABLE, then definition' ordering.  It keeps everything together neatly.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Natural.hs:46]: This actually falls into the 'use partial function to ensure totality of another function', so it should be OK. You can also use the monus operator `^-` instead if you prefer.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Natural.hs:74]: Slightly awkward wording here: better choice would be '... returns a tuple: its first element is the prefix of xs of length n, ...'

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Ord.hs:7]: Same here - you're definitely the maintainer of this one.

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Ord.hs:30]: This is Boolean blind; a better choice for the function argument is `a -> a -> Ordering`.

File: plutus-list/test/Spec.hs
Comment (kozross) [inline plutus-list/test/Spec.hs:57]: I would actually wrap these generators and shrinkers by using newtypes, with `Arbitrary` instances over them.

**Review by kozross — CHANGES_REQUESTED** [review]

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (kozross) [inline plutus-list/src/PlutusTx/List/Natural.hs:32]: The style guide requires [explicit foralls](https://github.com/mlabs-haskell/styleguide/blob/main/TLDR.md#type-and-kind-signatures): this is kind of necessary if we're using `ScopedTypeVariables`.

**Comment by kozross** [comment]
You can write `ordNub` similarly to a mergesort, except that the merging step removes adjacent equal elements.

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (re-xyr) [reply plutus-list/src/PlutusTx/List/Ord.hs:30]: If we use `Ordering` then we can't implement `isSortedMonotonic(On)` in terms of this; we'll need to define another `isSortedMonotonicBy`. Not sure if this is good.

Also, I'm not sure if passing in a function returning `Ordering` is less blind. It actually can confuse user more b/c nowhere in the signature suggests if the function checks for ascending or descending order. By passing a `Bool`-returning function instead there isn't such confusion. For example, I'd argue it is more clear to write `isSortedBy (>=)` compared to ``isSortedBy (compare `on` Down)``.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (re-xyr) [reply plutus-list/src/PlutusTx/List/Natural.hs:74]: I just copied the words from the haddock of `base` - will change this.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (re-xyr) [reply plutus-list/src/PlutusTx/List/Natural.hs:46]: I wasn't aware of the monus operator, will change this.

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (kozross) [reply plutus-list/src/PlutusTx/List/Ord.hs:30]: Fair point on both, let's leave it as-is then.

File: plutus-list/src/PlutusTx/List/Natural.hs
Comment (jeffhappily) [reply plutus-list/src/PlutusTx/List/Natural.hs:26]: @kozross I think probably it's worth mentioning this in the style guide as well so we can keep things consistent

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (re-xyr) [inline plutus-list/src/PlutusTx/List/Ord.hs:125]: This definition is borrowed from `base`.

File: plutus-list/src/PlutusTx/List/Ord.hs
Comment (re-xyr) [inline plutus-list/src/PlutusTx/List/Ord.hs:140]: Not sure why `fourmolu` complains about not finding `BangPatterns` even if it is enabled in the cabal file.

**Comment by re-xyr** [comment]
It seems that nix build is failing. Is there something wrong with the current Action workflow?

**Comment by kozross** [comment]
@re-xyr There is - it just got fixed on `master`, so once you merge, the issue should go away.

**Comment by re-xyr** [comment]
The current sizes of the functions are:

```
Size checks
  Natural
    length fits on-chain:              OK
      Size: 121B
    take fits on-chain:                OK
      Size: 173B
    drop fits on-chain:                OK
      Size: 168B
    splitAt fits on-chain:             OK
      Size: 263B (~0.5KiB)
    replicate fits on-chain:           OK
      Size: 163B
  Ord
    sort fits on-chain:                OK
      Size: 747B (~0.5KiB)
    sortOn fits on-chain:              OK
      Size: 762B (~0.5KiB)
    sortBy fits on-chain:              OK
      Size: 547B (~0.5KiB)
    ordNub fits on-chain:              OK
      Size: 821B (~1KiB)
    ordNubBy fits on-chain:            OK
      Size: 620B (~0.5KiB)
    isSorted fits on-chain:            OK
      Size: 426B (~0.5KiB)
    isSortedOn fits on-chain:          OK
      Size: 428B (~0.5KiB)
    isSortedAscending fits on-chain:   OK
      Size: 426B (~0.5KiB)
    isSortedAscendingOn fits on-chain: OK
      Size: 428B (~0.5KiB)
    isSortedBy fits on-chain:          OK
      Size: 218B
```

We can slice ~200B off `sortBy` (as well as `sortOn`, `sort`, `ordNub`, `ordNubBy`) by using a smaller (and somewhat more inefficient) implementation of mergesort than the definition in `base`:

```
  Ord
    sort fits on-chain:                OK
      Size: 533B (~0.5KiB)
    sortOn fits on-chain:              OK
      Size: 547B (~0.5KiB)
    sortBy fits on-chain:              OK
      Size: 332B (~0.5KiB)
    ordNub fits on-chain:              OK
      Size: 606B (~0.5KiB)
    ordNubBy fits on-chain:            OK
      Size: 406B (~0.5KiB)
```

Is this worth it?

Also I'm not sure why `formolu` complains about `BangPatterns` in CI even with the extension enabled.

**Comment by kozross** [comment]
Anything to reduce size is worthwhile, mostly because Plutus inlines basically everything. We're willing to accept some inefficiencies in this case: 280B is a huge improvement, so go ahead and make that change.

**Review by kozross — APPROVED** [review]

**Pull request merged**
